### ISSUE TITLE
Tracking progress

### ISSUE THREAD

**Issue by xiaoguang1988** [metaOmic/metaOmics#24]
Is it possible for users to know their progress?
E.g. in metaClust, when tuning K and W, user could be aware of their permutation progress. Is it possible to pipe R print message to shiny webpage? If so, how could I prepare my R code.
Currently I am piping out the progress by cat or print.

**Comment by CHPGenetics** [comment]
I'm not sure how to pass a dynamic value inside a function out of the function. We can discuss this after we finished our main works

**Comment by schwannden** [comment]
Currently there is no good solution except writing your own Shiny-progress compatible plot function.
* tutorial: http://shiny.rstudio.com/articles/progress.html
* example: http://shiny.rstudio.com/gallery/progress-bar-example.html

**Comment by xiaoguang1988** [comment]
You have messenger.css
Is it possible that you wrap up the message and pop to a new messenger?

A stupid way is write stop(MSG.any), then the message will be conducted to error messenger.
Not sure if this is valid or correct.

**Comment by schwannden** [comment]
The thing is, the way you output progress in your TuneK function is through outputting the message to console. And while our Shiny is running your TuneK function, there is no way of collecting your function output. Even if we redirect the output to a file, we can't read it until the TuneK is finished and exit. This is the blocking nature of Shiny's application, this is a bad design, but a more strait forward and easier understood one.

To sum up, there are two main challenges:
1. How do we get the progress output, this could possibly be solved by redirecting the outpour to a file.
2. How do we get the output while the TuneK function is still executing? I Don't see any way to do it.

**Comment by schwannden** [comment]
stop(MSG.any) is not a viable solution, as it completely stops the execution. Yes I could catch the error and send the message to browser, but your execution has stopped! Of course now I can catch the error, output to browser, and restart the execution with additional information on where to resume. But this way, your function need to be able to be called with an additional information about past information and where to resume. Which is exactly the same as re-writing the whole function so that upon each iteration, it notifies Shiny through progress function.

**Comment by xiaoguang1988** [comment]
I see.
This is just an enhancement and won't bother too much.
May be we can leave it for future.

**Comment by schwannden** [comment]
I believe @CHPGenetics is happy to hear that XD

**Comment by xiaoguang1988** [comment]
XD

**Comment by schwannden** [comment]
Let's close it for now. It does't seems like we have time for this in any near future
